libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs

```diff
@@ -156,42 +160,85 @@ public string Evaluate(string inlineStr, object scope = null, IGetMethod methodB
             return evaluator.EvaluateTemplate(fakeTemplateId, scope);
         }
 
-        private void ImportIds(string[] ids, Dictionary<string, LGResource> sources, ImportResolverDelegate importResolver)
+        /// <summary>
+        /// Parse lg content to LGResource list.
+        /// All the imports of the content will be also tracked down to parse.
+        /// </summary>
+        /// <param name="content">Text content contains lg templates.</param>
+        /// <param name="name">Text name.</param>
+        /// <param name="importResolver">resolver to resolve LG import id to template text.</param>
+        /// <returns>LGTemplate list of parsed lg content.</returns>
+        private List<LGResource> ParseContent(string content, string name, ImportResolverDelegate importResolver)
+        {
+            var rootResource = LGParser.Parse(content, name);
+            var parsedResources = this.DiscoverLGResources(rootResource, importResolver);
+
+            return parsedResources;
+        }
+
+        /// <summary>
+        /// Discover all imported lg resources from a start resouce.
+        /// </summary>
+        /// <param name="start">The lg resource from which to start discovering imported resources.</param>
+        /// <param name="importResolver">resolver to resolve LG import id to template text.</param>
+        /// <returns>LGResource list of parsed lg content.</returns>
+        private List<LGResource> DiscoverLGResources(LGResource start, ImportResolverDelegate importResolver)
         {
+            var resourceIdsFound = new HashSet<string>();
+
             if (importResolver == null)
             {
                 // default to fileResolver...
                 importResolver = FileResolver;
             }
 
-            foreach (var id in ids)
+            return ResolveImportResources(start, importResolver, resourceIdsFound);
+        }
+
+        /// <summary>
+        /// Resolve imported LG resources from a start resource.
+        /// All the imports will be visited and resolved to LGResouce list.
+        /// </summary>
+        /// <param name="start">The lg resource from which to start resolving imported resources.</param>
+        /// <param name="importResolver">resolver to resolve LG import id to template text.</param>
+        /// <param name="resourceIdsFound">Resource ids that have been vistied.</param>
+        /// <returns>LGTemplate list of parsed lg content.</returns>
+        private List<LGResource> ResolveImportResources(LGResource start, ImportResolverDelegate importResolver, HashSet<string> resourceIdsFound)
```

Comment:
Can this just be HashSet<LGResource>?

Reply:
You mean the return value should be HashSet<LGResource>?

Reply:
fixed